sockeye/loss.py

```diff
@@ -168,87 +163,3 @@ def update(self, labels, preds):
                 # NOTE: This is different from MXNet's metrics
                 self.num_inst += batch_size
             self.sum_metric += mx.nd.sum(loss).asscalar()
-
-
-class SmoothedCrossEntropyLoss(Loss):
-    """
-    Computes a smoothed cross-entropy loss. Smoothing is defined by alpha which indicates the
-    amount of probability mass subtracted from the true label probability (1-alpha).
-    Alpha is then uniformly distributed across other labels.
-    If normalize is True, the gradient by dividing by the number of non-PAD tokens.
-
-    :param loss_config: Loss configuration.
-    """
-
-    def __init__(self, loss_config: LossConfig) -> None:
-        self.loss_config = loss_config
-        utils.check_condition(self.loss_config.smoothed_cross_entropy_alpha >= 0,
-                              "alpha for smoothed loss must be >= 0")
-
-    def get_loss(self, logits: mx.sym.Symbol, labels: mx.sym.Symbol) -> List[mx.sym.Symbol]:
-        """
-        Returns loss and softmax output symbols given logits and integer-coded labels.
-
-        :param logits: Shape: (batch_size * target_seq_len, target_vocab_size).
-        :param labels: Shape: (batch_size * target_seq_len,).
-        :return: List of loss and softmax output symbols.
-        """
-        probs = mx.sym.softmax(data=logits)
-
-        on_value = 1.0 - self.loss_config.smoothed_cross_entropy_alpha
-        off_value = self.loss_config.smoothed_cross_entropy_alpha / (self.loss_config.vocab_size - 1.0)
-        cross_entropy = mx.sym.one_hot(indices=mx.sym.cast(data=labels, dtype='int32'),
-                                       depth=self.loss_config.vocab_size,
-                                       on_value=on_value,
-                                       off_value=off_value)
-
-        # zero out pad symbols (0)
-        cross_entropy = mx.sym.where(labels, cross_entropy, mx.sym.zeros((0, self.loss_config.vocab_size)))
-
-        # compute cross_entropy
-        cross_entropy = cross_entropy * - mx.sym.log(data=probs + 1e-10)
-        cross_entropy = mx.sym.sum(data=cross_entropy, axis=1)
-
-        if self.loss_config.normalize:
-            cross_entropy = _normalize(cross_entropy, labels)
-
-        cross_entropy = mx.sym.MakeLoss(cross_entropy, name=C.SMOOTHED_CROSS_ENTROPY)
-        probs = mx.sym.BlockGrad(probs, name=C.SOFTMAX_NAME)
-        return [cross_entropy, probs]
-
-    def create_metric(self) -> "SmoothedCrossEntropyMetric":
-        return SmoothedCrossEntropyMetric(self.loss_config)
-
-
-class SmoothedCrossEntropyMetric(EvalMetric):
```

Comment:
It's great that we can merge smoothed and non-smoothed loss!  We should make sure that `CrossEntropyMetric` can also apply smoothing to match `CrossEntropyLoss`.

Reply:
I thought about that, but its also actually not too bad to have the same loss values for smoothed and non-smoothed gradients for easy comparison. I will look into it.

Reply:
The main place we're using the loss-replicating metrics is for optimizers like Eve that want the exact loss from training.  The idea is that `Loss` classes can return corresponding `Metric` versions that use the same config and return the same results.  For checkpoints, we're still using the vanilla MXNet metrics.

Reply:
Ah you are right, I will put that back in then!

Reply:
I added back the smoothed cross-entropy for the Loss metric. However, this is super slow as we are not caching the label distributions as before in the symbol graph. This probably removes all the benefit we have from the fusioning of the smoothed CE grad in MXNet 0.12.
I will think about a different way for this.

Reply:
Do we have a way to return the loss?  The custom metrics were originally only used for the Eve optimizer, which just needs the loss.

Reply:
The problem is the smoothed loss in general. Explicitly creating the soft labels / label distribution is super expensive. As it is right now, we reduce speed by 60%.

Reply:
I will leave as is right now as this is correct. If we see that eve + label smoothing is too slow for us we may take another look at this.